### PR TITLE
Add a tutor plugin to easily run this on a devstack

### PR DESCRIPTION
**Pull request by bradenmacdonald**
It's a bit tricky to get this working on a Tutor-based devstack, but this plugin should make it easy!

This updates the README to explain how to use this MFE on a Tutor-based devstack, using the included plugin. See https://github.com/openedx/frontend-app-library-authoring/blob/bb9ee6590cc715daaf74fe84947ccc0baf3a07b8/tutor-contrib-library-authoring-mfe/README.rst for the details on how to use it.

Tested end-to-end using the drag and drop v2 XBlock. Other XBlocks may not work correctly for unrelated reasons.

CI failure is unrelated.

### CODE REVIEW

**Comment by openedx-webhooks** [comment]
Thanks for the pull request, @bradenmacdonald!  Please note that it may take us up to several weeks or months to complete a review and merge your PR.

Feel free to add as much of the following information to the ticket as you can:
- supporting documentation
- [Open edX discussion forum threads](https://discuss.openedx.org/)
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).


Please let us know once your PR is ready for our review and all tests are green.

**Comment by natabene** [comment]
@bradenmacdonald Thank you for your contribution. Please let me know once this is ready for our review.

**Comment by bradenmacdonald** [comment]
@natabene It's ready now. There is an error in the CI build but it's not related to this. I'm not sure if this repo has a maintainer at the moment? Because the build is failing because nobody is updating the dependencies of this repo and/or several related repos like @edx/frontend-component-footer

**Comment by connorhaugh** [comment]
Braden, I gave this a test (for local dev development) on my local tutor and it seems to work. I'll give a thumb here when we resolve the unrelated CI issues and see this PR merged.

**Comment by connorhaugh** [comment]
@bradenmacdonald a rebase off master should fix the CI issues.

**Comment by bradenmacdonald** [comment]
@connorhaugh Thanks!! Yep, that did it.

**Comment by brian-smith-tcril** [comment]
I tested this locally.

## tl;dr
I needed to run `pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_contrib_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe` for this to install correctly.

## Questions
* Is escaping the `&` is required on all platforms?
* What are the conventions around using `contrib` in tutor plugin names?
  * Should `tutor-contrib-library-authoring-mfe` be updated to be `tutor-library-authoring-mfe` in `setup.py` before this gets merged?

## Play by play

First issue I encountered was related to the line
```
pip install -e git+https://github.com/openedx/frontend-app-library-authoring.git#egg=tutor_library_authoring_mfe&subdirectory=tutor-contrib-library-authoring-mfe
```
which, for testing purposes, i updated to be
```
pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_library_authoring_mfe&subdirectory=tutor-contrib-library-authoring-mfe
```
when running it, I encountered an error
```
ERROR: tutor_library_authoring_mfe from git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_library_authoring_mfe (from -r /tmp/pipenv-ikn3q7vk-requirements/pipenv-a7yegfe3-requirement.txt (line 1)) does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
```
which I resolved by escaping the `&` by using `\&`
```
pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe
```
when running that, I encountered another error
```
  WARNING: Generating metadata for package tutor_library_authoring_mfe produced metadata for project name tutor-contrib-library-authoring-mfe. Fix your #egg=tutor_library_authoring_mfe fragments.
ERROR: Could not find a version that satisfies the requirement tutor-library-authoring-mfe (unavailable) (from versions: none)
ERROR: No matching distribution found for tutor-library-authoring-mfe (unavailable)
```
I'm not sure what the conventions around using `contrib` are, but that seemed to be causing the issue.

I pushed the branch to my fork, and made a commit (https://github.com/brian-smith-tcril/frontend-app-library-authoring/commit/49827046f6869f95b29a3f3b1e220e189844c7fd) updating `setup.py` to use `tutor-library-authoring-mfe` as the name instead of `tutor-contrib-library-authoring-mfe`

installing this using
```
pip install -e git+https://github.com/brian-smith-tcril/frontend-app-library-authoring.git@tutor-plugin#egg=tutor_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe
```
worked as expected, at which point I uninstalled and decided to try
```
pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_contrib_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe
```
which worked as expected

**Comment by brian-smith-tcril** [comment]
I tested this locally.

## tl;dr
I needed to run `pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_contrib_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe` for this to install correctly.

## Questions
* Is escaping the `&` is required on all platforms?
* What are the conventions around using `contrib` in tutor plugin names?
  * Should `tutor-contrib-library-authoring-mfe` be updated to be `tutor-library-authoring-mfe` in `setup.py` before this gets merged?

## Play by play

First issue I encountered was related to the line
```
pip install -e git+https://github.com/openedx/frontend-app-library-authoring.git#egg=tutor_library_authoring_mfe&subdirectory=tutor-contrib-library-authoring-mfe
```
which, for testing purposes, i updated to be
```
pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_library_authoring_mfe&subdirectory=tutor-contrib-library-authoring-mfe
```
when running it, I encountered an error
```
ERROR: tutor_library_authoring_mfe from git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_library_authoring_mfe (from -r /tmp/pipenv-ikn3q7vk-requirements/pipenv-a7yegfe3-requirement.txt (line 1)) does not appear to be a Python project: neither 'setup.py' nor 'pyproject.toml' found.
```
which I resolved by escaping the `&` by using `\&`
```
pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe
```
when running that, I encountered another error
```
  WARNING: Generating metadata for package tutor_library_authoring_mfe produced metadata for project name tutor-contrib-library-authoring-mfe. Fix your #egg=tutor_library_authoring_mfe fragments.
ERROR: Could not find a version that satisfies the requirement tutor-library-authoring-mfe (unavailable) (from versions: none)
ERROR: No matching distribution found for tutor-library-authoring-mfe (unavailable)
```
I'm not sure what the conventions around using `contrib` are, but that seemed to be causing the issue.

I pushed the branch to my fork, and made a commit (https://github.com/brian-smith-tcril/frontend-app-library-authoring/commit/49827046f6869f95b29a3f3b1e220e189844c7fd) updating `setup.py` to use `tutor-library-authoring-mfe` as the name instead of `tutor-contrib-library-authoring-mfe`

installing this using
```
pip install -e git+https://github.com/brian-smith-tcril/frontend-app-library-authoring.git@tutor-plugin#egg=tutor_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe
```
worked as expected, at which point I uninstalled and decided to try
```
pip install -e git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor_contrib_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe
```
which worked as expected

**Comment by brian-smith-tcril** [comment]
After getting it running, I tried creating a content library and adding some components. Creating the library worked as expected, but every component is failing to load with the error: `Cannot GET /library_authoring/xblock-bootstrap.html`

I'm thinking this may be related to the change from `library_authoring` to `library-authoring` in this commit https://github.com/openedx/frontend-app-library-authoring/commit/d719044cb8f570ba719736aac76fef914e5b4b6c

File: tutor-contrib-library-authoring-mfe/README.rst [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+Installation and usage
+----------------------
+
+**IMPORTANT**: First, you must be using `Tutor Nightly <https://docs.tutor.overhang.io/tutorials/nightly.html>`_, a
+recent (July 2022 or newer) version of edx-platform (`master` branch), and the
+`Tutor MFE plugin <https://github.com/overhangio/tutor-mfe/>`_.
+
+Then, follow these instructions to enable this microfrontend:
+
+1. Install this plugin: ``pip install -e git+https://github.com/openedx/frontend-app-library-authoring.git#egg=tutor_library_authoring_mfe&subdirectory=tutor-contrib-library-authoring-mfe``
```
Comment (brian-smith-tcril) [inline tutor-contrib-library-authoring-mfe/README.rst:16]:
```suggestion
1. Install this plugin: ``pip install -e git+https://github.com/openedx/frontend-app-library-authoring.git#egg=tutor_library_authoring_mfe\&subdirectory=tutor-contrib-library-authoring-mfe``
```
i ran into issues with the `&` not being escaped when running this locally

File: tutor-contrib-library-authoring-mfe/tutor_library_authoring_mfe/plugin.py [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+# Tell the tutor-mfe plugin about this MFE so we can build, run, and use it:
+hooks.Filters.CONFIG_DEFAULTS.add_item(("LIBRARY_AUTHORING_MFE_APP", {
+    "name": "library-authoring",
+    "repository": "https://github.com/openedx/frontend-app-library-authoring",
+    "port": 3001,
+    "env": {
+        "development": {
+            # Tutor requires the name of the MFE in the URL so this file won't be found at /xblock-bootstrap.html,
+            # which is the default location for development.
+            "SECURE_ORIGIN_XBLOCK_BOOTSTRAP_HTML_URL": "/library_authoring/xblock-bootstrap.html",
```
Comment (brian-smith-tcril) [inline tutor-contrib-library-authoring-mfe/tutor_library_authoring_mfe/plugin.py:39]:
```suggestion
            "SECURE_ORIGIN_XBLOCK_BOOTSTRAP_HTML_URL": "/library-authoring/xblock-bootstrap.html",
```
tested this change on my fork and it resolved the `Cannot GET /library_authoring/xblock-bootstrap.html` error I encountered.

**Review by brian-smith-tcril — CHANGES_REQUESTED** [review]
Overall this is really great! I'm super excited to see MFEs working as tutor plugins!

The big one here is the change in `plugin.py`, without that this just doesn't work.

I'm not sure about escaping the `&` in the readme, but running the command didn't work for me without it. I'm also not sure what to do about the `tutor_contrib_library_authoring_mfe` vs `tutor_library_authoring_mfe` thing, so it'd be helpful to get some outside input on that.

Looking forward to seeing what we can do to get this merged in!

**Comment by arbrandes** [comment]
Thanks for the in-depth review, @brian-smith-tcril!  From where I stand those are all great catches, though I'll let @bradenmacdonald weigh in and/or pull in your changes before we merge.

Also,

> I'm not sure what the conventions around using `contrib` are

There's some prior art by our colleague @kdmccormick: 

https://github.com/openedx/tutor-contrib-coursegraph

Which is to say, I'd try to fix it in the _other_ direction, if possible.  (In other words, getting the rid of the error when the package is prefixed with `tutor-contrib`.)

**Comment by kdmccormick** [comment]
I'm excited about this plugin @bradenmacdonald, and thanks for working through that @brian-smith-tcril. 

I agree that the Python distribution name (the thing defined in `setup.py`) should be **tutor-contrib-library-authoring-mfe**. That way, if this is ever pushed to PyPI, it would be installed as `pip install tutor-contrib-library-authoring-mfe`, clarifying that it's a contributor plugin, not an official one.

The Python package name is good to remain as `tutor_library_authoring_mfe`. That'll ensure that the plugin can be enabled via `tutor enable library_authoring_mfe` instead of the awkward `tutor enable contrib_library_authoring_mfe`.

(the docs on all of this ^ aren't great yet, but they're deductible by reverse engineering https://github.com/overhangio/cookiecutter-tutor-plugin)

Some other notes:

* The `#egg=` should match the Python distribution name, not the Python package name. That is, it should be `#egg=tutor-contrib-library-authoring-mfe`, not `#egg=tutor_library_authoring_mfe`. I think it is working somewhat by coincidence currently because pip is willing to silently translate the underscores into hyphens.
* Instead of escaping the `&`, we could wrap the entire `git+https://...` thing in single quotes. That way, bash won't treat `&` as a special character.

**Comment by bradenmacdonald** [comment]
Thanks for the review @brian-smith-tcril ! I've pushed bd5f2318687c9ad57f9186e189ecd491cc12989d to make those fixes. Would you mind checking if that did it?

> What are the conventions around using contrib in tutor plugin names?

Using `contrib` is required - Regis has asked that any plugins that aren't created by him, i.e. any unofficial plugins, use `contrib` in the name to clearly distinguish them from plugins supported by overhang.io.

**Review by brian-smith-tcril — APPROVED** [review]
Perfect!

I can verify this worked locally for me when I followed the instructions in `tutor-contrib-library-authoring-mfe/README.rst` when changing the install step to
```
pip install -e 'git+https://github.com/open-craft/frontend-app-library-authoring.git@tutor-devstack-plugin#egg=tutor-contrib-library-authoring-mfe&subdirectory=tutor-contrib-library-authoring-mfe'
```
I can verify again using the actual line from the README after this is merged.

**Comment by bradenmacdonald** [comment]
@brian-smith-tcril @arbrandes Can we merge this now?

**Comment by openedx-webhooks** [comment]
@bradenmacdonald 🎉 Your pull request was merged!  Please take a moment  to [answer a two question survey](https://docs.google.com/forms/d/e/1FAIpQLSceJOyGJ6JOzfy6lyR3T7EW_71OWUnNQXp68Fymsk3MkNoSDg/viewform?usp=pp_url&entry.1671973413=openedx/frontend-app-library-authoring&entry.867055334=https://github.com/openedx/frontend-app-library-authoring/pull/50&entry.1484655318=https://github.com/bradenmacdonald&entry.752974735=2022-07-05+04:55&entry.1917517419=2022-10-17+19:51&entry.2133058324=Yes) so we can improve your experience in the future.

**Comment by arbrandes** [comment]
Let's merge this!  Apologies for the delay, @bradenmacdonald, and thanks for the contrib!
